Compilation fails on MinGw 6.2.0 x64

Protobuf for CPP v3.0.0 release
MinGW64  6.2.0.
Cmake 3.7.0.
Windows Server 2008 R2.
Build cmake generated proj. fails.
```
........
[ 61%] Building CXX object CMakeFiles/tests.dir/D_/Libraries/Unpacked/protobuf/protobuf-cpp-3.0.0/protobuf-3.0.0/src/google/protobuf/unittest_no_arena_lite.pb.cc.obj
c:/mingw-w64/x86_64-6.2.0-posix-sjlj-rt_v5-rev1/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/6.2.0/../../../../x86_64-w64-mingw32/bin/as.exe: CMakeFiles/tests.dir/D_/Libraries/Unpacked/protobuf/protobuf-cpp-3.0.0/protobuf-3.0.0/src/google/protobuf/unittest_well_known_types.pb.cc.obj: too many sections (33968)
C:\Users\admin\AppData\Local\Temp\2\cc57VLqh.s: Assembler messages:
C:\Users\admin\AppData\Local\Temp\2\cc57VLqh.s: Fatal error: can't write CMakeFiles/tests.dir/D_/Libraries/Unpacked/protobuf/protobuf-cpp-3.0.0/protobuf-3.0.0/src/google/protobuf/unittest_well_known_types.pb.cc.obj: File too big
c:/mingw-w64/x86_64-6.2.0-posix-sjlj-rt_v5-rev1/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/6.2.0/../../../../x86_64-w64-mingw32/bin/as.exe: CMakeFiles/tests.dir/D_/Libraries/Unpacked/protobuf/protobuf-cpp-3.0.0/protobuf-3.0.0/src/google/protobuf/unittest_well_known_types.pb.cc.obj: too many sections (33968)
C:\Users\admin\AppData\Local\Temp\2\cc57VLqh.s: Fatal error: can't close CMakeFiles/tests.dir/D_/Libraries/Unpacked/protobuf/protobuf-cpp-3.0.0/protobuf-3.0.0/src/google/protobuf/unittest_well_known_types.pb.cc.obj: File too big
make[2]: *** [CMakeFiles/tests.dir/D_/Libraries/Unpacked/protobuf/protobuf-cpp-3.0.0/protobuf-3.0.0/src/google/protobuf/unittest_well_known_types.pb.cc.obj] Error 1
make[2]: *** Waiting for unfinished jobs....
c:/mingw-w64/x86_64-6.2.0-posix-sjlj-rt_v5-rev1/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/6.2.0/../../../../x86_64-w64-mingw32/bin/as.exe: CMakeFiles/tests.dir/D_/Libraries/Unpacked/protobuf/protobuf-cpp-3.0.0/protobuf-3.0.0/src/google/protobuf/map_unittest.pb.cc.obj: too many sections (34918)
C:\Users\admin\AppData\Local\Temp\2\cctglMC3.s: Assembler messages:
C:\Users\admin\AppData\Local\Temp\2\cctglMC3.s: Fatal error: can't write CMakeFiles/tests.dir/D_/Libraries/Unpacked/protobuf/protobuf-cpp-3.0.0/protobuf-3.0.0/src/google/protobuf/map_unittest.pb.cc.obj: File too big
c:/mingw-w64/x86_64-6.2.0-posix-sjlj-rt_v5-rev1/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/6.2.0/../../../../x86_64-w64-mingw32/bin/as.exe: CMakeFiles/tests.dir/D_/Libraries/Unpacked/protobuf/protobuf-cpp-3.0.0/protobuf-3.0.0/src/google/protobuf/map_unittest.pb.cc.obj: too many sections (34918)
C:\Users\admin\AppData\Local\Temp\2\cctglMC3.s: Fatal error: can't close CMakeFiles/tests.dir/D_/Libraries/Unpacked/protobuf/protobuf-cpp-3.0.0/protobuf-3.0.0/src/google/protobuf/map_unittest.pb.cc.obj: File too big
make[2]: *** [CMakeFiles/tests.dir/D_/Libraries/Unpacked/protobuf/protobuf-cpp-3.0.0/protobuf-3.0.0/src/google/protobuf/map_unittest.pb.cc.obj] Error 1
make[1]: *** [CMakeFiles/tests.dir/all] Error 2
make: *** [all] Error 2

```

--- a/cmake/tests.cmake
+++ b/cmake/tests.cmake
@@ -206,6 +206,13 @@ endif()
 
 if(MINGW)
   set_source_files_properties(${tests_files} PROPERTIES COMPILE_FLAGS "-Wno-narrowing")
+
+  # required for tests on MinGW Win64
+  if (CMAKE_SIZEOF_VOID_P EQUAL 8)
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,--stack,16777216")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wa,-mbig-obj")
+  endif()
+
 endif()
 
 add_executable(tests ${tests_files} ${common_test_files} ${tests_proto_files} ${lite_test_proto_files})

